File: backend/src/auth/auth.service.ts
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { JwtService } from '@nestjs/jwt';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { hashPassword, verifyHash } from '../common/hash';
import { UserEntity } from '../database/entities';
import { ChangePasswordDto } from './dto/change-password.dto';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(UserEntity) private readonly users: Repository<UserEntity>,
    private readonly jwt: JwtService,
    private readonly config: ConfigService,
  ) {}

  async getAccessToken(username: string, password: string): Promise<string> {
    const user = await this.users.findOne({ where: { username } });
    if (!user || !(await verifyHash(password, user.passwordHash))) {
      throw new HttpException({ error: 'wrong username or password' }, HttpStatus.BAD_REQUEST);
    }

    return this.jwt.sign(
      { sub: username },
      {
        secret: this.config.get<string>('JWT_SECRET') ?? 'secret',
        expiresIn: '720h',
      },
    );
  }

  async changePassword(username: string, request: ChangePasswordDto): Promise<void> {
    if (request.newPassword !== request.repeatNewPassword) {
      throw new HttpException({ error: 'New password not equal to repeated one' }, HttpStatus.BAD_REQUEST);
    }

    const user = await this.users.findOne({ where: { username } });
    if (!user || !(await verifyHash(request.password, user.passwordHash))) {
      throw new HttpException({ error: 'wrong username or password' }, HttpStatus.BAD_REQUEST);
    }

    user.passwordHash = await hashPassword(request.newPassword);
    await this.users.save(user);
  }
}

File: backend/services/auth_service.go
package services

import (
	"errors"
	"keef/common"
	"keef/database"
	"keef/models"
	"os"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

var jwtSecret = []byte(os.Getenv("JWT_SECRET"))

func GetAccessToken(username string, password string) (accessToken string, err error) {
	var user database.UserEntity
	lookupErr := database.DB.Where("username = ?", username).First(&user).Error
	if lookupErr != nil || !common.VerifyHash(password, user.PasswordHash) { // Wrong username or password
		return "", errors.New("wrong username or password")
	}

	expirationTime := time.Now().Add(720 * time.Hour)
	claims := &jwt.RegisteredClaims{
		ExpiresAt: jwt.NewNumericDate(expirationTime),
		Subject:   username,
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenString, err := token.SignedString(jwtSecret)
	if err != nil {
		return "", err
	}

	return tokenString, nil
}

func ChangePassword(username string, request models.ChangePasswordRequest) (err error) {
	if request.NewPassword != request.RepeatNewPassword {
		return errors.New("New password not equal to repeated one")
	}

	var user database.UserEntity
	lookupErr := database.DB.Where("username = ?", username).First(&user).Error
	if lookupErr != nil || !common.VerifyHash(request.Password, user.PasswordHash) { // Wrong username or password
		return errors.New("wrong username or password")
	}

	hash, error := common.Hash(request.NewPassword)
	if error != nil {
		return errors.New("Something went wrong")
	}

	database.DB.Model(&user).Update("password_hash", hash)

	return nil
}
